### PR TITLE
More accurate control of create before destroy behaviors

### LINKED ISSUE #34: aws_security_group_rule create_before_destroy triggers bug in provider
## Describe the Bug

This module creates Security Group Rules using `create_before_destroy = true`. 
This causes Terraform to fail when adding or removing CIDRs to an existing rule where an existing CIDR is retained, due to an issue with the Terraform AWS provider.

See https://github.com/hashicorp/terraform-provider-aws/issues/25173 for details and examples. 

See also https://github.com/hashicorp/terraform/issues/31316 for proposed solutions.

### PR DESCRIPTION
**Pull request by Nuru**
## note

README and code are (hopefully) final.

This will be released as v2.0.0-rc1 due to changed defaults, requirement for Terraform 1.0, and possible service interruption when upgrading. Migration document TBD.

## what

- Make `create_before_destroy` default to `true` for security groups 
- Introduce `preserve_security_group_id` to control replacement of security group when rules change

## why

- In most cases, replacing a security group with `create_before_destroy = false` will fail with an error because you cannot destroy a security group while it is associated with a resource
- Fixes #34 

## references

- Fixes #34 
- https://github.com/hashicorp/terraform-provider-aws/issues/25173
- https://github.com/hashicorp/terraform/issues/31316

### CODE REVIEW

File: main.tf [hunk 1]
```diff
@@ -17,6 +17,12 @@ locals {
     local.create_security_group ? local.created_security_group.id : coalesce(var.target_security_group_id[0],
     "var.target_security_group_id contains null value. Omit value if you want this module to create a security group.")
   ) : null
+
+  # Setting `create_before_destroy` on the security group rules forces create_before_destroy behavior
+  # on the security group, so we have to disable it on the rules if disabled on the security group.
+  rule_create_before_destroy = var.create_before_destroy && var.rule_create_before_destroy
+  # We also have to make it clear to Terraform that the CBD rules will never reference the DBC security group
```
Comment (aknysh) [inline main.tf:24]:
```suggestion
  # We also have to make it clear to Terraform that the CBD rules will never reference the CBD security group
```

Comment (aknysh) [reply main.tf:24]: is `DBC` correct and means "Destroy before create"?
If yes, maybe update the comment to mention that, and also it's not clear at the first glance how the line below  makes it clear to Terraform that the CBD rules will never reference the xxx security group

File: variables.tf [hunk 2]
```diff
@@ -39,15 +39,32 @@ variable "security_group_description" {
...
     Note that changing this value will always cause the security group to be replaced.
     EOT
 }
 
+variable "rule_create_before_destroy" {
+  type        = bool
+  default     = true
+  description = <<-EOT
+    Set `true` to enable terraform `create_before_destroy` behavior on the created security group rules.
+    When `true`, a new security group can replace an existing security group with zero service interruption,
```
Comment (aknysh) [inline variables.tf:60]:
```suggestion
    When `true`, a new security group rule can replace an existing security group rule with zero service interruption,
```

**Review by aknysh — COMMENTED** [review]
please see comments

File: main.tf (hunk 1 above)
Comment (Nuru) [reply main.tf:30]: Comments updated

**Review by osterman — DISMISSED** [review]

**Comment by osterman** [comment]
Great depiction of the problem at a high-level. This is a great addition to the README.

**Pull request merged**